go-controller/pkg/ovn/pods_retry.go

```diff
@@ -0,0 +1,170 @@
+package ovn
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// iterateRetryPods checks if any outstanding pods have been waiting for 60 seconds of last known failure
+// then tries to re-add them if so
+// updateAll forces all pods to be attempted to be retried regardless of the 1 minute delay
+func (oc *Controller) iterateRetryPods(updateAll bool) {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	now := time.Now()
+	for key, podEntry := range oc.retryPods {
+		if podEntry.ignore && podEntry.logicalPortCacheEntry == nil {
+			// neither addition nor deletion is being retried
+			continue
+		}
+
+		pod := podEntry.pod
+		podDesc := fmt.Sprintf("[%s/%s/%s]", pod.UID, pod.Namespace, pod.Name)
+		// it could be that the Pod got deleted, but Pod's DeleteFunc has not been called yet, so don't retry creation
+		retryAdd := true
+		kPod, err := oc.watchFactory.GetPod(pod.Namespace, pod.Name)
+		if err != nil && errors.IsNotFound(err) {
+			klog.Infof("%s pod not found in the informers cache, not going to retry pod setup", podDesc)
+			retryAdd = false
+		}
+
+		if !util.PodScheduled(kPod) {
+			klog.V(5).Infof("Retry: %s not scheduled", podDesc)
+			continue
+		}
+		podEntry.backoffSec = (podEntry.backoffSec * 2)
+		if podEntry.backoffSec > 60 {
+			podEntry.backoffSec = 60
+		}
+		backoff := (podEntry.backoffSec * time.Second) + (time.Duration(rand.Intn(500)) * time.Millisecond)
+		podTimer := podEntry.timeStamp.Add(backoff)
+		if updateAll || now.After(podTimer) {
+			// check if we need to retry delete first
+			if podEntry.logicalPortCacheEntry != nil {
+				klog.Infof("%s retry pod teardown", podDesc)
+				if err := oc.deleteLogicalPort(pod); err == nil {
+					klog.Infof("%s pod teardown successful", podDesc)
+					podEntry.logicalPortCacheEntry = nil
+				} else {
+					klog.Infof("%s teardown retry failed; will try again later", podDesc)
+					podEntry.timeStamp = time.Now()
+					continue // if deletion failed we will not retry add
+				}
+			}
+			// check if we need to retry add
+			if retryAdd && !podEntry.ignore {
+				klog.Infof("%s retry pod setup", podDesc)
+				if err := oc.ensurePod(nil, kPod, true); err == nil {
+					klog.Infof("%s pod setup successful", podDesc)
+					delete(oc.retryPods, key) // this means retryDelete and retryAdd both worked, we can remove entry safely
+				} else {
+					klog.Infof("%s setup retry failed; will try again later", podDesc)
+					podEntry.timeStamp = time.Now()
+				}
+			}
```

Comment:
if we didnt need to retryAdd because the pod was gone, and we deleted correctly, shouldnt we now remove the entry from oc.retryPods?

Reply:
oh good point...yes you are right.

Reply:
done.